src/evaluators/source-4-3.js

```diff
@@ -0,0 +1,808 @@
+    list("<", (x, y) => x < y),
+    list("<=", (x, y) => x <= y),
+    list(">", (x, y) => x > y),
+    list(">=", (x, y) => x >= y),
+    list("!", x => !x)
+);
+
+// the global environment also has bindings for all
+// primitive non-function values, such as undefined and
+// math_PI
+
+const primitive_constants = list(
+    list("undefined", undefined),
+    list("math_PI", math_PI)
+);
+
+// setup_environment makes an environment that has
+// one single frame, and adds a binding of all names
+// listed as primitive_functions and primitive_values.
+// The values of primitive functions are "primitive"
+// objects, see line 281 how such functions are applied
+
+function setup_environment() {
+    const primitive_function_names =
+        map(f => head(f), primitive_functions);
+    const primitive_function_values =
+        map(f => make_primitive_function(head(tail(f))),
+            primitive_functions);
+    const primitive_constant_names =
+        map(f => head(f), primitive_constants);
+    const primitive_constant_values =
+        map(f => head(tail(f)),
+            primitive_constants);
+    return extend_environment(
+        append(primitive_function_names,
+            primitive_constant_names),
+        append(primitive_function_values,
+            primitive_constant_values),
+        the_empty_environment);
+}
+
+const the_global_environment = setup_environment();
+
+// below are amb related
+function is_amb(stmt) {
+    return is_tagged_list(stmt, "application") &&
+        is_name(operator(stmt)) &&
+        name_of_name(operator(stmt)) === "amb";
+}
+
+function amb_choices(stmt) {
+    return operands(stmt);
+}
+
+function ambeval(exp, env, succeed, fail) {
+    return analyze(exp)(env, succeed, fail);
+}
+
+function analyze_self_evaluating(stmt) {
+    return (env, succeed, fail) => succeed(stmt, fail);
+}
+
+function analyze_name(stmt) {
+    return (env, succeed, fail) =>
+        succeed(lookup_name_value(name_of_name(stmt), env),
+            fail);
+}
+
+function analyze_function_definition(stmt) {
+    const vars =
+        function_definition_parameters(stmt);
+    const bfun =
+        analyze(function_definition_body(stmt));
+    const locals = local_names(function_definition_body(stmt));
+    return (env, succeed, fail) =>
+        succeed(make_compound_function(vars, bfun, locals, env),
+            fail);
+}
+
+function analyze_conditional_expression(stmt) {
+    const pfun = analyze(cond_expr_pred(stmt));
+    const cfun = analyze(cond_expr_cons(stmt));
+    const afun = analyze(cond_expr_alt(stmt));
+    return (env, succeed, fail) =>
+        pfun(env,
+            // success continuation for evaluating the predicate
+            // to obtain pred_value
+            (pred_value, fail2) =>
+                is_true(pred_value) ?
+                    cfun(env, succeed, fail2) :
+                    afun(env, succeed, fail2),
+            fail);
+}
+
+function analyze_sequence(stmts) {
```

Comment:
What happens when return occurs in the middle of a sequence?

Reply:
The statements after the return statement should not be analysed. This was fixed in 
a15a077